File: src/main/java/com/hazza/LeetCodeByJava/stack/No84LargestRectangleInHistogram.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import com.hazza.LeetCodeByJava.datastructure.SegmentTree;
import org.omg.CORBA.MARSHAL;

import java.util.Stack;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/largest-rectangle-in-histogram/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-09-07
 * Time: 6:04 PM
 */
public class No84LargestRectangleInHistogram {

    // solution 1
    public int largestRectangleArea(int[] heights) {
        if (heights == null || heights.length == 0) return 0;
        int[] st = SegmentTree.constructST(heights);
        return getAreaRec(heights, st, 0, heights.length - 1);
    }

    private int getAreaRec(int[] heights, int[] st, int left, int right) {
        if (left > right) return Integer.MIN_VALUE;
        if (left == right) return heights[left];
        int m = SegmentTree.RMQ(heights, st, left, right);
        return Math.max(Math.max(getAreaRec(heights, st, left, m - 1),
                getAreaRec(heights, st, m + 1, right)),
                (right - left + 1) * heights[m]);
    }

    // solution 2
    public int largestRectangleArea1(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0, n = heights.length;

        for (int i = 0; i <= n ; i++) {
            int top = (i == n ? 0 : heights[i]);
            if (stack.isEmpty() || heights[stack.peek()] <= top) stack.push(i);
            else {
                top = stack.pop();
                maxArea = Math.max(maxArea,
                        heights[top] * (stack.isEmpty() ? i : i - stack.peek() - 1));
                --i;
            }
        }

        return maxArea;
    }
}

File: src/main/python/com/hazza/leetcode_by_python/stack/No84_Largest_Rectangle_in_Histogram.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-22
# https://leetcode.com/problems/largest-rectangle-in-histogram/
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        stack = [-1]

        heights.append(0)
        for i in range(len(heights)):
            while heights[i] < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i - stack[-1] - 1
                res = max(res, w * h)
            stack.append(i)
        heights.pop()

        return res


if __name__ == '__main__':
    solution = Solution()

    assert solution.largestRectangleArea([2, 1, 5, 6, 2, 3]) == 10
